src/kubeobjects/istio.go

```diff
@@ -0,0 +1,237 @@
+package kubeobjects
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net"
+
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1"
+	"github.com/Dynatrace/dynatrace-operator/src/dtclient"
+	"github.com/go-logr/logr"
+	istiov1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	istioclientset "istio.io/client-go/pkg/clientset/versioned"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type probeResult int
+
+const (
+	probeObjectFound probeResult = iota
+	probeObjectNotFound
+	ProbeTypeFound
+	ProbeTypeNotFound
+	probeUnknown
+)
+
+var (
+	IstioGVRName = "networking.istio.io"
+
+	// VirtualServiceGVK => definition of virtual service GVK for oneagent
+	VirtualServiceGVK = schema.GroupVersionKind{
+		Group:   IstioGVRName,
+		Version: "v1alpha3",
+		Kind:    "VirtualService",
+	}
+
+	// ServiceEntryGVK => definition of virtual service GVK for oneagent
+	ServiceEntryGVK = schema.GroupVersionKind{
+		Group:   IstioGVRName,
+		Version: "v1alpha3",
+		Kind:    "ServiceEntry",
+	}
+)
+
+// CheckIstioEnabled checks if Istio is installed
+func CheckIstioEnabled(cfg *rest.Config) (bool, error) {
+	discoveryclient, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return false, err
+	}
+	apiGroupList, err := discoveryclient.ServerGroups()
+	if err != nil {
+		return false, err
+	}
+
+	for _, apiGroup := range apiGroupList.Groups {
+		if apiGroup.Name == IstioGVRName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// BuildNameForEndpoint returns a name to be used as a base to identify Istio objects.
+func BuildNameForEndpoint(name string, protocol string, host string, port uint32) string {
+	sum := sha256.Sum256([]byte(fmt.Sprintf("%s-%s-%s-%d", name, protocol, host, port)))
+	return hex.EncodeToString(sum[:])
+}
+
+func buildObjectMeta(name string, namespace string) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      name,
+		Namespace: namespace,
+	}
+}
+
+func BuildIstioLabels(name, role string) map[string]string {
+	return map[string]string{
+		"dynatrace":            "oneagent",
+		"oneagent":             name,
+		"dynatrace-istio-role": role,
+	}
+}
+
+func isIp(host string) bool {
+	return net.ParseIP(host) != nil
+}
+
+func VerifyIstioCrdAvailability(instance *dynatracev1beta1.DynaKube, config *rest.Config) probeResult {
```

Comment:
could be in a `validate.go` under the `istio` package, as its used for validating both in the reconcile and in the validation webhook